app/src/org/commcare/engine/cases/CaseUtils.java

```diff
@@ -151,9 +126,80 @@ public static void purgeCases() throws InvalidCaseGraphException {
 
     }
 
+    private static Vector<String> getAllOwners() {
+        Vector<String> users = new Vector<>();
+        Vector<String> owners = new Vector<>();
+        SqlStorage<User> userStorage = CommCareApplication.instance().getUserStorage(User.STORAGE_KEY, User.class);
+        for (IStorageIterator<User> userIterator = userStorage.iterate(); userIterator.hasMore(); ) {
+            String id = userIterator.nextRecord().getUniqueId();
+            owners.addElement(id);
+            users.addElement(id);
+        }
+
+        //Now add all of the relevant groups
+        //TODO: Wow. This is.... kind of megasketch
+        for (String userId : users) {
+            DataInstance instance = CommCareUtil.loadFixture("user-groups", userId);
+            if (instance == null) {
+                continue;
+            }
+            EvaluationContext ec = new EvaluationContext(instance);
+            for (TreeReference ref : ec.expandReference(
+                    XPathReference.getPathExpr("/groups/group/@id").getReference())) {
+                AbstractTreeElement<AbstractTreeElement> idelement = ec.resolveReference(ref);
+                if (idelement.getValue() != null) {
+                    owners.addElement(idelement.getValue().uncast().getString());
+                }
+            }
+        }
+        return owners;
+    }
+
+    /**
+     * Get records ids for all related cases for the given set of cases including the original set of cases
+     *
+     * @param recordIds databse ids for the cases we want to find related cases
+     * @return database ids for all related cases for the given set of cases
+     */
+    public static Set<String> getRelatedCases(Set<String> recordIds) {
+        if (recordIds.isEmpty()) {
+            return recordIds;
+        }
+
+        SqlStorage<ACase> storage = CommCareApplication.instance().getUserStorage(ACase.STORAGE_KEY, ACase.class);
+        DAG<String, int[], String> fullCaseGraph = getFullCaseGraph(storage, new AndroidCaseIndexTable(),
+                new Vector<>());
+        Set<String> caseIds = getCaseIdsFromRecordsIds(storage, recordIds);
+        Set<String> relatedCaseIds = fullCaseGraph.getRelatedRecords(caseIds);
+        Set<String> relatedRecordIds = new HashSet<>();
+        for (String relatedCaseId : relatedCaseIds) {
+            String relatedRecordId = getRecordIdFromCaseId(storage, relatedCaseId);
```

Comment:
When this set of ID's is large (which it might be fairly often) it will be very expensive to be doing these reads one-by-one.

There are a few options for how to do this faster. **One** of the methods above that does a trawl of the full sets of case data must have been able to retain this mapping, so just keeping it around might be enough, but otherwise doing a [bulk read](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/services/storage/IStorageUtilityIndexed.java#L280) type operator is a much more efficient option.

Reply:
https://github.com/dimagi/commcare-android/pull/2955/commits/23836b694221cbc92fa5c894e4eec429c5fc6d1c